From 60cc4ea769da158d20a40fc04a2eb9b7d6ce78f7 Mon Sep 17 00:00:00 2001
From: Shahed Shaikh <shahedx.shaikh@intel.com>
Date: Mon, 29 Jun 2020 10:24:04 +0800
Subject: [PATCH] DRV_DPM_SW-294: DPM remove inst array from
 g_bridge_id_entry_hash_table

- no need to have inst specific entries for  g_bridge_id_entry_hash_table

Signed-off-by: Shahed Shaikh <shahedx.shaikh@intel.com>
---
 drivers/net/datapath/dpm/datapath.h       |  3 +--
 drivers/net/datapath/dpm/datapath_proc.c  |  4 +--
 drivers/net/datapath/dpm/datapath_swdev.c | 45 ++++++++++++-------------------
 drivers/net/datapath/dpm/datapath_swdev.h |  1 -
 4 files changed, 20 insertions(+), 33 deletions(-)

diff --git a/drivers/net/datapath/dpm/datapath.h b/drivers/net/datapath/dpm/datapath.h
index cce45c5e8ade..f5dfee3771e6 100644
--- a/drivers/net/datapath/dpm/datapath.h
+++ b/drivers/net/datapath/dpm/datapath.h
@@ -1019,8 +1019,7 @@ int32_t dp_get_netif_subifid_priv(int inst, struct net_device *netif,
 				  struct sk_buff *skb, void *subif_data,
 				  u8 dst_mac[DP_MAX_ETH_ALEN],
 				  dp_subif_t *subif, uint32_t flags);
-struct hlist_head *get_dp_g_bridge_id_entry_hash_table_info(int instance,
-							    int index);
+struct hlist_head *get_dp_g_bridge_id_entry_hash_table_info(int index);
 struct dp_evt_notif_info {
 	struct notifier_block nb;
 	struct dp_event evt_info;
diff --git a/drivers/net/datapath/dpm/datapath_proc.c b/drivers/net/datapath/dpm/datapath_proc.c
index 0b81fc658057..b540075645f6 100644
--- a/drivers/net/datapath/dpm/datapath_proc.c
+++ b/drivers/net/datapath/dpm/datapath_proc.c
@@ -59,7 +59,7 @@ static int proc_brvlan_start(void)
 	br_hash_index = 0;
 
 	brdev_info = hlist_entry_safe
-		     ((get_dp_g_bridge_id_entry_hash_table_info(0,
+		     ((get_dp_g_bridge_id_entry_hash_table_info(
 				     br_hash_index))->first,
 		      struct br_info, br_hlist);
 	return 0;
@@ -79,7 +79,7 @@ static int proc_brvlan_dump(struct seq_file *s, int pos)
 			return -1;
 		}
 		brdev_info = hlist_entry_safe
-			     ((get_dp_g_bridge_id_entry_hash_table_info(0,
+			     ((get_dp_g_bridge_id_entry_hash_table_info(
 					     br_hash_index))->first,
 			      struct br_info, br_hlist);
 	}
diff --git a/drivers/net/datapath/dpm/datapath_swdev.c b/drivers/net/datapath/dpm/datapath_swdev.c
index 1fdf6c8b6a9c..e1cc19e84f33 100644
--- a/drivers/net/datapath/dpm/datapath_swdev.c
+++ b/drivers/net/datapath/dpm/datapath_swdev.c
@@ -20,8 +20,7 @@ static int dp_swdev_add_bport_to_list(struct br_info *br_item,
 				      struct bridge_id_entry_item *brid_item);
 static int dp_swdev_del_bport_from_list(struct br_info *br_item, int bport);
 
-struct hlist_head
-	g_bridge_id_entry_hash_table[DP_MAX_INST][BR_ID_ENTRY_HASH_TABLE_SIZE];
+struct hlist_head g_bridge_id_entry_hash_table[BR_ID_ENTRY_HASH_TABLE_SIZE];
 
 static inline u32 dp_swdev_cal_hash(const void *ptr)
 {
@@ -36,11 +35,10 @@ void dp_kfree(void const *ptr)
 	devm_kfree(&g_dp_dev->dev, (void *)ptr);
 }
 
-struct hlist_head *get_dp_g_bridge_id_entry_hash_table_info(int inst, int idx)
+struct hlist_head *get_dp_g_bridge_id_entry_hash_table_info(int idx)
 {
-	return &g_bridge_id_entry_hash_table[inst][idx];
+	return &g_bridge_id_entry_hash_table[idx];
 }
-EXPORT_SYMBOL(get_dp_g_bridge_id_entry_hash_table_info);
 
 int dp_get_fid_by_dev(struct net_device *dev, int *inst)
 {
@@ -113,19 +111,16 @@ struct br_info *dp_swdev_bridge_entry_lookup_rcu(struct net_device *dev)
 	struct br_info *br_item;
 	struct hlist_head *head;
 	u32 idx;
-	int i;
 
 	idx = dp_swdev_cal_hash(dev);
 
-	for (i = 0; i < DP_MAX_INST; i++) {
-		head = (&g_bridge_id_entry_hash_table[i][idx]);
-		hlist_for_each_entry_rcu(br_item, head, br_hlist) {
-			if (br_item->dev == dev) {
-				DP_DEBUG(DP_DBG_FLAG_SWDEV,
-					 "Hash entry found(%s)\n",
-					 dev->name);
-				return br_item;
-			}
+	head = get_dp_g_bridge_id_entry_hash_table_info(idx);
+	hlist_for_each_entry_rcu(br_item, head, br_hlist) {
+		if (br_item->dev == dev) {
+			DP_DEBUG(DP_DBG_FLAG_SWDEV,
+				 "Hash entry found(%s)\n",
+				 dev->name);
+			return br_item;
 		}
 	}
 
@@ -152,7 +147,7 @@ static void dp_swdev_insert_bridge_id_entry(struct br_info *br_item)
 
 	idx = dp_swdev_cal_hash(br_item->dev);
 
-	head = &g_bridge_id_entry_hash_table[br_item->inst][idx];
+	head = get_dp_g_bridge_id_entry_hash_table_info(idx);
 
 	hlist_add_head_rcu(&br_item->br_hlist, head);
 
@@ -160,7 +155,7 @@ static void dp_swdev_insert_bridge_id_entry(struct br_info *br_item)
 		 idx, br_item->dev->name);
 }
 
-int dp_swdev_bridge_id_entry_free(int instance)
+void dp_swdev_bridge_id_entry_free(void)
 {
 	struct hlist_node *node;
 	struct hlist_head *head;
@@ -168,22 +163,19 @@ int dp_swdev_bridge_id_entry_free(int instance)
 	u32 idx;
 
 	for (idx = 0; idx < BR_ID_ENTRY_HASH_TABLE_SIZE; idx++) {
-		head = &g_bridge_id_entry_hash_table[instance][idx];
+		head = get_dp_g_bridge_id_entry_hash_table_info(idx);
 		hlist_for_each_entry_safe(br_item, node, head, br_hlist) {
 			dp_swdev_remove_bridge_id_entry(br_item);
 		}
 	}
-
-	return DP_SUCCESS;
 }
 
 int dp_swdev_bridge_id_entry_init(void)
 {
-	int i, j;
+	int i;
 
-	for (i = 0; i < DP_MAX_INST; i++)
-		for (j = 0; j < BR_ID_ENTRY_HASH_TABLE_SIZE; j++)
-			INIT_HLIST_HEAD(&g_bridge_id_entry_hash_table[i][j]);
+	for (i = 0; i < BR_ID_ENTRY_HASH_TABLE_SIZE; i++)
+		INIT_HLIST_HEAD(&g_bridge_id_entry_hash_table[i]);
 
 	return DP_SUCCESS;
 }
@@ -1140,10 +1132,7 @@ int dp_ndo_bridge_dellink(struct net_device *dev, struct nlmsghdr *nlh,
 
 void dp_switchdev_exit(void)
 {
-	int i;
-
-	for (i = 0; i < DP_MAX_INST; i++)
-		dp_swdev_bridge_id_entry_free(i);
+	dp_swdev_bridge_id_entry_free();
 }
 
 int dp_switchdev_init(void)
diff --git a/drivers/net/datapath/dpm/datapath_swdev.h b/drivers/net/datapath/dpm/datapath_swdev.h
index 5ee40a3c4517..1b0ca6f36869 100644
--- a/drivers/net/datapath/dpm/datapath_swdev.h
+++ b/drivers/net/datapath/dpm/datapath_swdev.h
@@ -77,7 +77,6 @@ struct fdb_tbl {
 };
 
 extern struct list_head fdb_tbl_list;
-int dp_swdev_bridge_id_entry_free(int instance);
 int dp_swdev_bridge_id_entry_init(void);
 struct br_info *dp_swdev_bridge_entry_lookup_rcu(struct net_device *dev);
 int dp_swdev_chk_bport_in_br(struct net_device *bp_dev, int bport,
